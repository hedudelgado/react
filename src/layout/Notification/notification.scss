@import "../../css/main";

.pi-notification {
  position: relative;
  padding: 10px 10px 10px 41px;
  color: $pi-notification-black;
  font-size: $font-size--s;

  .icon {
    font-size: 1.5rem;
    position: absolute;
    left: 10px;
    top: 5px;
  }

  .title {
    display: block;
    font-weight: bold;
    font-size: $font-size--s;
    margin-top: -2px;
  }

  a, a:hover, a:visited, a:active {
    color: $pi-notification-black;
    text-decoration: underline;
  }

  &.priority-1 {
    background-color: $pi-notification-error-light;

    .icon {
      display: flex;
      justify-content: center;
      padding-top: 0;
      background-color: $pi-notification-error-dark;

      &:before {
        align-self: center;
        font-family: 'wb-icons';
        font-size: 2rem;
        color: $pi-notification-white;
        content: '\e95f';
      }
    }

    .error-link {
      color: $pi-notification-error-dark;
      text-decoration: underline;
      font-weight: bold;
      font-size: $font-size--base;
    }
  }

  &.priority-2 {
    &-warning {
      background-color: $pi-notification-warning-light;

      .icon {
        background-color: $pi-notification-warning-dark;
      }
    }

    &-error {
      background-color: $pi-notification-error-light;

      .icon {
        background-color: $pi-notification-error-dark;
      }
    }

    &-info {
      background-color: $pi-notification-info-light;

      .icon {
        background-color: $pi-notification-info-dark;
      }
    }

    &-success {
      background-color: $pi-notification-success-light;

      .icon {
        background-color: $pi-notification-success-dark;
      }
    }

    .icon {
      padding-top: 5px;
    }
  }

  &.priority-1, &.priority-2 {
    display: flex;
    padding: 0;

    .content {
      align-self: center;
      padding: 10px;
    }

    .title {
      margin-top: 0;
    }

    .icon {
      position: relative;
      top: 0;
      left: 0;
      text-align: center;
      flex-basis: 50px;
      min-width: 50px;
      font-size: $font-size--xl;
      color: $pi-notification-white;
    }
  }

  &.priority-2, &.priority-3 {
    &-warning {
      background-color: $pi-notification-warning-light;
    }

    &-error {
      background-color: $pi-notification-error-light;
    }

    &-info {
      background-color: $pi-notification-info-light;
    }

    &-success {
      background-color: $pi-notification-success-light;
    }
  }

  &.priority-4 {
    &-warning {
      border: 1px solid $pi-notification-warning-dark;
    }

    &-error {
      border: 1px solid $pi-notification-error-dark;
    }

    &-info {
      border: 1px solid $pi-notification-info-dark;
    }

    &-success {
      border: 1px solid $pi-notification-success-dark;
    }
  }

  &.priority-3, &.priority-4, &.priority-5 {
    &-warning {
      .icon {
        color: $pi-notification-warning-dark;
      }
    }

    &-error {
      .icon {
        color: $pi-notification-error-dark;
      }
    }

    &-info {
      .icon {
        color: $pi-notification-info-dark;
      }
    }

    &-success {
      .icon {
        color: $pi-notification-success-dark;
      }
    }
  }

  &.sticky {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    z-index: 999;
  }
}
