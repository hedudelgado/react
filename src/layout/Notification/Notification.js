import React from 'react';
import {
  bool,
  node,
  oneOf,
  string,
} from 'prop-types';
import classNames from 'classnames';
import Icon from '../Icon';
import BaseStyle from '../BaseStyle';
import Typography from '../Typography';
import './notification.scss';

const variants = {
  warning: {
    className: 'warning',
    icon: 'disc-info-new',
  },
  error: {
    className: 'error',
    icon: 'disc-info-new',
  },
  success: {
    className: 'success',
    icon: 'tick',
  },
  info: {
    className: 'info',
    icon: 'disc-info-new',
  },
};

const Notification = ({
  priority,
  variant,
  ariaLive,
  show,
  sticky,
  title,
  children,
  className,
  ...otherProps
}) => {
  const classes = classNames(
    'pi-notification',
    `priority-${priority}`,
    `priority-${priority}-${variants[variant].className}`,
    (sticky && 'sticky'),
    className,
  );

  return show && children && (
    <BaseStyle component="div" role="region" aria-live={ariaLive} className={classes} {...otherProps}>
      <Icon icon={variants[variant].icon} className="icon" />
      <div className="content">
        {title && <Typography component="span" className="title">{title}</Typography>}
        {children}
      </div>
    </BaseStyle>
  );
};

Notification.propTypes = {
  /**
   * Priority defines the style of the Notification based on importance.
   * 1 is the highest priority and 5 the lowest.
   * Priority 1 supports only Errors.
   * Tip: `.error-link` className in a Priority 1's child anchor tag or button, will style it
   * like in the example.
   */
  priority: oneOf([1, 2, 3, 4, 5]).isRequired,
  /**
   * Variant defines the Notification's purpose.
   */
  variant: oneOf(['warning', 'error', 'success', 'info']).isRequired,
  /**
   * Aria-Live is used to set the priority with which a screen reader should treat
   * updates to live regions.
   * The possible settings are: off, polite or assertive.
   */
  ariaLive: oneOf(['polite', 'assertive', 'off']).isRequired,
  /**
   * A boolean value which displays or hides the Notification.
   */
  show: bool,
  /**
   * A boolean value which which pins the Notification to top of the page.
   */
  sticky: bool,
  /**
   * An optional title in the Notification.
   */
  title: string,
  /**
   * The Notification contents. It is usually a text, but it can also be further DOM.
   */
  children: node,
};

Notification.defaultProps = {
  show: false,
  sticky: false,
  title: '',
  children: null,
};

export default Notification;
