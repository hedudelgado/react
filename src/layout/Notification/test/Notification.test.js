import React from 'react';
import { shallow } from 'enzyme';
import Notification from '../Notification';

describe('Notification component', () => {
  const priorities = [1];
  const variants = ['warning'];

  const combinedProps = [];
  priorities.forEach((priority) => {
    variants.forEach((variant) => {
      combinedProps.push([priority, variant]);
    });
  });

  it('should show the component', () => {
    const props = {
      ariaLive: 'polite',
      variant: 'warning',
      priority: 2,
      show: true,
    };
    const wrapper = shallow(<Notification {...props}>This is a notification</Notification>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a title if one is present', () => {
    const props = {
      ariaLive: 'polite',
      variant: 'warning',
      priority: 2,
      show: true,
      title: 'Some title',
    };
    const wrapper = shallow(<Notification {...props}>This is a notification</Notification>);

    expect(wrapper).toContainExactlyOneMatchingElement('.title');
  });

  it('should render a sticky notification', () => {
    const props = {
      ariaLive: 'polite',
      variant: 'warning',
      priority: 2,
      show: true,
      sticky: true,
    };
    const wrapper = shallow(<Notification {...props}>This is a notification</Notification>);

    expect(wrapper).toContainExactlyOneMatchingElement('.sticky');
  });

  it('should hide the component', () => {
    const props = {
      ariaLive: 'polite',
      variant: 'warning',
      priority: 2,
      show: false,
      sticky: true,
    };
    const wrapper = shallow(<Notification {...props}>This is a notification</Notification>);

    expect(wrapper).toBeEmptyRender();
  });

  it.each(combinedProps)(
    'should render Notification with priority %i and variant %s',
    (priority, variant) => {
      const wrapper = shallow(<Notification ariaLive="off" priority={priority} variant={variant} show>Text</Notification>);
      expect(wrapper).toHaveClassName(`pi-notification priority-${priority} priority-${priority}-${variant}`);
    },
  );
});
