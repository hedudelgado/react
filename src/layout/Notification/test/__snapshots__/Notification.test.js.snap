// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component should show the component 1`] = `
<BaseStyle
  aria-live="polite"
  className="pi-notification priority-2 priority-2-warning"
  component="div"
  role="region"
>
  <Icon
    className="icon"
    icon="disc-info-new"
  />
  <div
    className="content"
  >
    This is a notification
  </div>
</BaseStyle>
`;
